### PR TITLE
Segment Anything Model example with 🤗 Transformers

### PR DESCRIPTION
**Pull request by merveenoyan**
This PR includes md, notebook and script for SAM example, built together with @sayakpaul

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks for your PR. 

As a rule of this repository, let's only add the markdown file and not anything else. When the PR is approved, we can add the rest of the files.

File: examples/vision/sam.py [hunk 1]
```diff
@@ -0,0 +1,601 @@
+"""
+Title: Segment Anything Model with 🤗Transformers
+Author: [Sayak Paul](https://twitter.com/RisingSayak) & [Merve Noyan](https://twitter.com/mervenoyann)
```
Comment (sayakpaul) [inline examples/vision/sam.py:3]:
```suggestion
Authors: [Merve Noyan](https://twitter.com/mervenoyann) & [Sayak Paul](https://twitter.com/RisingSayak)
```

You have put in a lot of effort in this example. You should be the primary author.

File: examples/vision/sam.py [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+Date created: 2023/07/11
+Last modified: 2023/07/11
+Description: Implementing Segment Anything Model using 🤗Transformers.
```
Comment (sayakpaul) [inline examples/vision/sam.py:6]:
```suggestion
Description: Fine-tuning Segment Anything Model using Keras and 🤗 Transformers.

File: examples/vision/sam.py [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+
+In the [Segment Anything Model](https://segment-anything.com/) (dubbed as SAM),
+researchers from Meta extended the space of language prompting to visual prompting. SAM
+is capable of performing zero-shot segmentation with a prompt input, inspired by large
+language models. The prompt here can be a set of foreground/background points, free text,
+a box or a mask. There are many downstream segmentation tasks, including semantic
+segmentation and edge detection. The goal of SAM is to enable all of these downstream
+segmentation tasks through prompting.
+
+In this notebook, we'll learn how to use the SAM model from 🤗 Transformers for performing
```
Comment (sayakpaul) [inline examples/vision/sam.py:35]:
```suggestion
In this example, we'll learn how to use the SAM model from 🤗 Transformers for performing
```

File: examples/vision/sam.py [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+"""
+
+"""
+## SAM in a few words
+
+SAM has the following components:
+
+|
+![](https://scontent.fccu31-1.fna.fbcdn.net/v/t39.2365-6/338558258_1349701259095991_4358060436604292355_n.png?_nc_cat=104&ccb=1-7&_nc_sid=ad8a9d&_nc_ohc=f0B0G6TtZG0AX-2gBP1&_nc_ht=scontent.fccu31-1.fna&oh=00_AfDSPCChiBRx_ZmE_krAlqsfSIZU3BUmRJcZIc4LVZa93w&oe=64737F89)
+![](https://scontent.fccu31-1.fna.fbcdn.net/v/t39.2365-6/338558258_1349701259095991_4358060436604292355_n.png?_nc_cat=104&ccb=1-7&_nc_sid=ad8a9d&_nc_ohc=f0B0G6TtZG0AX-2gBP1&_nc_ht=scontent.fccu31-1.fna&oh=00_AfDSPCChiBRx_ZmE_krAlqsfSIZU3BUmRJcZIc4LVZa93w&oe=64737F89)
```
Comment (sayakpaul) [inline examples/vision/sam.py:56]: Seems like this one is extra

Comment (sayakpaul) [reply examples/vision/sam.py:56]: Also, this URL tends to behave weirdly. Let's use: https://i.imgur.com/oLfdwuB.png.

File: examples/vision/sam.py [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+from transformers import TFSamModel, SamProcessor
+
+model = TFSamModel.from_pretrained("facebook/sam-vit-base")
+processor = SamProcessor.from_pretrained("facebook/sam-vit-base")
+
+"""
+Next, we write some utility functions for visualization. Most of these functions are
+taken from [this
+notebook](https://github.com/huggingface/notebooks/blob/main/examples/segment_anything.ipynb).
+notebook](https://github.com/huggingface/notebooks/blob/main/examples/segment_anything.ipynb).
```
Comment (sayakpaul) [inline examples/vision/sam.py:113]: Needs fixing.

File: examples/vision/sam.py [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+
+remote_path = "https://hf.co/datasets/sayakpaul/sample-datasets/resolve/main/breast-cancer-dataset.tar.gz"
+dataset_path = keras.utils.get_file(
+    "breast-cancer-dataset.tar.gz", remote_path, untar=True
+)
+
+"""
+Let's now visualize a sample from the dataset.
+
+*(The `show_mask()` utility is taken from [this
```
Comment (sayakpaul) [inline examples/vision/sam.py:323]: The links should stay on a single line. So, in this case, we should do:

```
*(The `show_mask()` utility is taken from
[this notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/SAM/Fine_tune_SAM_(segment_anything)_on_a_custom_dataset.ipynb)
)*
```

Notice how the link is on the same line. You can follow [this source Python file](https://github.com/keras-team/keras-io/blob/master/examples/nlp/mlm_training_tpus.py) to understand this idea.

File: examples/vision/sam.py [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+sample = next(iter(train_ds))
+for k in sample:
+    print(k, sample[k].shape, sample[k].dtype, isinstance(sample[k], tf.Tensor))
+
+"""
+### Training
+"""
+
+"""
+We will now write DICE loss. This implementation is based on [MONAI DICE
```
Comment (sayakpaul) [inline examples/vision/sam.py:485]: Same for the links.

File: examples/vision/sam.py (hunk 6 above)
Comment (merveenoyan) [reply examples/vision/sam.py:323]: I remember removing some of them, it's nb2py script that does it :/

Comment (sayakpaul) [reply examples/vision/sam.py:323]: Yes, so, we need to manually remove them.

File: examples/vision/sam.py (hunk 1 above)
Comment (ariG23498) [reply examples/vision/sam.py]: I have mad respect for this gesture. Who else but @sayakpaul!

File: examples/vision/sam.py [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+"""
+## SAM in a few words
+
+SAM has the following components:
+
+|
+![](https://imgur.com/oLfdwuB)
+|:--:|
+| Image taken from the official [SAM blog
+post](https://ai.facebook.com/blog/segment-anything-foundation-model-image-segmentation/)
```
Comment (ariG23498) [inline examples/vision/sam.py:58]:
```suggestion
| Image taken from the official [SAM blog post](https://ai.facebook.com/blog/segment-anything-foundation-model-image-segmentation/) |
```

I have seen that links should be in a single line.

File: examples/vision/sam.py [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+"""
+
+"""
+SAM was pre-trained to predict a _valid_ mask for any acceptable prompt. This requirement
+allows SAM to output a valid mask even when the prompt is ambiguous to understand -- this
+makes SAM ambiguity-aware. Moreover, SAM predicts multiple masks for a single prompt.
+
+We highly encourage you to check out the [SAM paper](https://arxiv.org/abs/2304.02643)
+and the [blog
+post](https://ai.facebook.com/blog/segment-anything-foundation-model-image-segmentation/)
```
Comment (ariG23498) [inline examples/vision/sam.py:85]:
```suggestion
and the [blog post](https://ai.facebook.com/blog/segment-anything-foundation-model-image-segmentation/)
```

File: examples/vision/sam.py [hunk 10]
```diff
@@ -0,0 +1,599 @@
...
+
+There are three checkpoints for SAM:
+
+* [sam-vit-base](https://huggingface.co/facebook/sam-vit-base)
+* [sam-vit-large](https://huggingface.co/facebook/sam-vit-large)
+* [sam-vit-huge](https://huggingface.co/facebook/sam-vit-huge).
+
+We load `sam-vit-base` in
+[`TFSamModel`](https://huggingface.co/docs/transformers/main/model_doc/sam#transformers.TF
+SamModel). We also need `SamProcessor`for the associated checkpoint.
```
Comment (ariG23498) [inline examples/vision/sam.py:100]:
```suggestion
[`TFSamModel`](https://huggingface.co/docs/transformers/main/model_doc/sam#transformers.TFSamModel). We also need `SamProcessor`for the associated checkpoint.
```

File: examples/vision/sam.py [hunk 11]
```diff
@@ -0,0 +1,599 @@
...
+
+from transformers import TFSamModel, SamProcessor
+
+model = TFSamModel.from_pretrained("facebook/sam-vit-base")
+processor = SamProcessor.from_pretrained("facebook/sam-vit-base")
+
+"""
+Next, we write some utility functions for visualization. Most of these functions are
+taken from [this
+notebook](https://github.com/huggingface/notebooks/blob/main/examples/segment_anything.ipynb).
```
Comment (ariG23498) [inline examples/vision/sam.py:111]:
```suggestion
taken from [this notebook](https://github.com/huggingface/notebooks/blob/main/examples/segment_anything.ipynb).
```

File: examples/vision/sam.py [hunk 12]
```diff
@@ -0,0 +1,599 @@
...
+tf.shape(ground_truth_seg)
+
+"""
+### Preparing `tf.data.Dataset`
+
+We now write a generator class to prepare the images and the segmentation masks using the
+`processor` utilized above. We will leverage this generator class to create a
+`tf.data.Dataset` object for our training set by using
+`tf.data.Dataset.from_generator()`. Utilities of this class have been adapted from [this
+notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/SAM/Fine_tune_SAM_(segment_anything)_on_a_custom_dataset.ipynb).
```
Comment (ariG23498) [inline examples/vision/sam.py:367]:
```suggestion
`tf.data.Dataset.from_generator()`. Utilities of this class have been adapted from
[this notebook](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/SAM/Fine_tune_SAM_(segment_anything)_on_a_custom_dataset.ipynb).
```

**Review by ariG23498 — COMMENTED** [review]
This is so cool!

Most of my comments are on the link format.

File: examples/vision/sam.py [hunk 13]
```diff
@@ -0,0 +1,595 @@
...
+"""
+### Training
+"""
+
+"""
+We will now write DICE loss. This implementation is based on [MONAI DICE
+loss](https://docs.monai.io/en/stable/_modules/monai/losses/dice.html#DiceLoss).
+"""
+
+from tensorflow.keras import backend as K
```
Comment (fchollet) [inline examples/vision/sam.py:483]: Don't use this namespace, it's deprecated. Instead use tf ops like `tf.reduce_sum`

File: examples/vision/sam.py [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
+
+
+"""
+## Fine-tuning SAM
+
+We will now fine-tune SAM's decoder part. We will freeze the vision encoder and prompt
+encoder layers.
+"""
+
+from tensorflow import keras
```
Comment (fchollet) [inline examples/vision/sam.py:514]: Move all imports to the top

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by merveenoyan** [comment]
@fchollet I addressed your comments. Thanks a lot for reviewing this.

**Comment by merveenoyan** [comment]
@fchollet `autogen.py add_example vision/sam.py` ran seamlessly on my local. Can someone from your team re-run the GHAction that failed?

**Review by fchollet — COMMENTED** [review]
The PR is missing the md and ipynb files, please commit them.

**Comment by merveenoyan** [comment]
@fchollet Docker tests pass. even though I ran black before `autogen.py` it can't find anything to change, I checked the version of the black to match the one in GH Actions workflow (it's different from the one in `requirements.txt`file, I can open a PR to pin that one to match the one in GHActions if you want). I saw the formatting problem comes from another script, and someone else opened a PR for it here #1442.

**Comment by merveenoyan** [comment]
@fchollet since that PR for other example's black error is merged I synced my fork and my tests seem to pass, can you or someone from Keras team give another review?

**Comment by merveenoyan** [comment]
pinging @fchollet here

**Comment by pcoet** [comment]
@merveenoyan Thanks for addressing @fchollet's feedback. Can you also fix the links that @sayakpaul mentioned? They need to be on one line, as described [here](https://github.com/keras-team/keras-io/blob/master/contributor_guide.md#markdown-links). Then you'll also need to regenerate the .md and .ipynb files.

**Comment by merveenoyan** [comment]
@pcoet can you give me another review?

**Review by pcoet — APPROVED** [review]

**Comment by pcoet** [comment]
Thanks, @merveenoyan!

**Comment by sayakpaul** [comment]
@pcoet could you leave the URL of the tutorial when it's up on keras.io?
